### ISSUE TITLE
AssignPreviousVisits job fails when run asynchronously

### ISSUE THREAD

**Issue by JackWH** [kyranb/Footprints#48]
The `TrackRegistrationAttribution` trait has an `assignPreviousVisits()` method, which gets called when a `TrackableInterface` model's `created` event fires. This method dispatches the `AssignPreviousVisits` job — which is supposed to run on the queue if `config('footprints.async') === true`.

Unfortunately queuing the job fails every time, because:

> Serialization of 'Closure' is not allowed

Laravel can't serialize the `protected Request $request` property on the class, as the `Request` class uses closures.

The way `AssignPreviousVisits` is implemented requires backwards-incompatible changes to become queuable again. This is because the `Request` class has been type-hinted in several places. It's possible to work around this issue, but it's not very elegant... (cc @bilfeldt, in case I'm missing anything?).

1. Create a custom function on your `TrackableInterface` model implementation. A custom function is required, as you can't just override the existing `trackRegistration()` function without the model becoming incompatible with `TrackableInterface`:

```php
/**
 * Assign earlier visits using current request.
 *
 * @see \Kyranb\Footprints\TrackRegistrationAttribution::trackRegistration()
 */
public function trackRegistrationViaFootprint(string $footprint): void
{
    Visit::unassignedPreviousVisits($footprint)->update([
        config('footprints.column_name') => $this->id,
    ]);
}
```

2. Create a custom copy of the `AssignPreviousVisits` job, changing the `Request $request` property for `string $footprint`, and calling the new `trackRegistrationViaFootprint()` method:

```php
/**
 * @see \Kyranb\Footprints\Jobs\AssignPreviousVisits
 */
class AssignPreviousVisitsViaFootprint implements ShouldQueue
{
    use Dispatchable, InteractsWithQueue, Queueable, SerializesModels;

    protected string $footprint;
    protected TrackableInterface $trackable;

    public function __construct(string $footprint, TrackableInterface $trackable)
    {
        $this->footprint = $footprint;
        $this->trackable = $trackable;
    }

    public function handle()
    {
        $this->trackable->trackRegistrationViaFootprint($this->footprint);

        event(new RegistrationTracked($this->trackable));
    }
}
```

3. Back on the `TrackableInterface` model, override the `assignPreviousVisits()` function from the `TrackRegistrationAttribution` trait, to call the custom job with a `string $footprint` (which can be serialized):

```php
/**
 * Sync visits from the logged in user before they registered.
 *
 * @see \Kyranb\Footprints\TrackRegistrationAttribution::assignPreviousVisits()
 */
public function assignPreviousVisits(): void
{
    $job = new AssignPreviousVisitsViaFootprint(request()->footprint(), $this);

    if (config('footprints.async') === true) {
        dispatch($job);
    } else {
        $job->handle();
    }
}
```

**Comment by bilfeldt** [comment]
@JackWH Thanks for a well-explaned issue. Unfortunately, we did not have tests covering this.

Would you be open to PR'ing a test illustrating the issue you are stating above? Hopefully, my PR #49 fixes this, but it would be nice to test it correctly.

**Comment by bilfeldt** [comment]
@kyranb tests were implemented in #49 as well so this issue was closed with that PR.
